src/net/stream.rs

```diff
@@ -87,31 +98,72 @@ impl TcpStream {
             read_fut.set(async move { notify.notified().await });
 
             match Self::recv(self.pair, buf) {
-                Some(res) => return Poll::Ready(res),
+                Some(res) => {
+                    self.is_closed = res == 0;
+                    return Poll::Ready(Ok(()));
+                }
                 _ => ready!(read_fut.poll(cx)),
             }
         }
     }
 
-    fn recv(pair: SocketPair, buf: &mut ReadBuf<'_>) -> Option<io::Result<()>> {
+    fn recv(pair: SocketPair, buf: &mut ReadBuf<'_>) -> Option<usize> {
         match World::current(|world| world.recv_on(pair)) {
-            Some(seg) => match seg {
-                Segment::Data(bytes) => {
-                    buf.put_slice(bytes.as_ref());
-                    return Some(Ok(()));
-                }
-            },
+            Some(seg) => {
+                let bytes = match seg {
+                    Segment::Data(bytes) => bytes,
```

Comment:
What happens if a data frame of zero bytes will that trigger some other eof behavior?

Reply:
Good catch, that is broken. We should no-op the write side to make this impossible.

Reply:
See [d8a156b](https://github.com/tokio-rs/turmoil/pull/42/commits/d8a156b34d26c7e91ce5b9459a51afc8a3b42354) for fix and test.